File: comgongdaeng/DynamicProgramming/BOJ_9465.py
# 1 2 3
# 4 5 6
# dp[0][0] 1을 뜯었을 때, 다음 스티커로 5를 선택할 것인가? 6을 선택할 것인가?
# dp[1][0] 4를 뜯었을 때, 다음 스티커로 2를 선택할 것인가? 3을 선택할 것인가?
# max(dp[0][0], dp[1][0])
import sys
n = int(sys.stdin.readline())
for _ in range(n):
    col = int(sys.stdin.readline())
    dp = [list(map(int, sys.stdin.readline().split())) for _ in range(2)]
    if col == 1:
        print(max(dp[0][0],dp[1][0]))
    else:
        dp[0][1] += dp[1][0]
        dp[1][1] += dp[0][0]
        for i in range(2, col):
            dp[0][i] += max(dp[1][i-1], dp[1][i-2])
            dp[1][i] += max(dp[0][i-1], dp[0][i-2])
        print(max(dp[0][col-1], dp[1][col-1]))
# 저도 모르게 이 친구가 3주차 pr에 같이 딸려가 머지까지 된 것을 너무 늦게 알아서,,
# 20, 21번째 라인 주석 추가하여 다시 커밋합니다.. :)

File: keemsebin/DynamicProgramming/BOJ_9465.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
const T = Number(input[0]);

for (let i = 0; i < T; i++) {
  const case1 = Number(input[i * 3 + 1]);

  const score1 = input[i * 3 + 2].split(" ").map(Number);
  const score2 = input[i * 3 + 3].split(" ").map(Number);

  const dp = [[0], [...score1], [...score2]];

  for (let j = 1; j < case1; j++) {
    dp[0][j] = Math.max(dp[1][j - 1], dp[2][j - 1]); // 아무것도 선택하지 않았을 경우 직전 열 중 최댓값
    dp[1][j] = Math.max(dp[0][j - 1], dp[2][j - 1]) + score1[j];
    dp[2][j] = Math.max(dp[0][j - 1], dp[1][j - 1]) + score2[j];
  }
  console.log(Math.max(dp[0][case1 - 1], dp[1][case1 - 1], dp[2][case1 - 1]));
}
